Source/Core/Core/HW/WiimoteEmu/MotionPlus.cpp

```diff
@@ -0,0 +1,376 @@
+    // It even works when removing the is_mp_data bit in the last byte
+    // My M+ non-inside gives: 61,46,45,aa,0,2 or b6,46,45,9a,0,2
+    // static const u8 init_data[6] = {0x8e, 0xb0, 0x4f, 0x5a, 0xfc | 0x01, 0x02};
+    static const u8 init_data[6] = {0x81, 0x46, 0x46, 0xb6, 0x01, 0x02};
+    std::copy(std::begin(init_data), std::end(init_data), data);
+    reg_data.cert_ready = 0x2;
+    return;
+  }
+
+  if (0x2 == reg_data.cert_ready)
+  {
+    static const u8 init_data[6] = {0x7f, 0xcf, 0xdf, 0x8b, 0x4f, 0x82};
+    std::copy(std::begin(init_data), std::end(init_data), data);
+    reg_data.cert_ready = 0x8;
+    return;
+  }
+
+  if (0x8 == reg_data.cert_ready)
+  {
+    // A real wiimote takes about 2 seconds to reach this state:
+    reg_data.cert_ready = 0xe;
+  }
+
+  if (0x18 == reg_data.cert_ready)
+  {
+    // TODO: determine the meaning of this
+    const u8 mp_cert2[64] = {
+        0xa5, 0x84, 0x1f, 0xd6, 0xbd, 0xdc, 0x7a, 0x4c, 0xf3, 0xc0, 0x24, 0xe0, 0x92,
+        0xef, 0x19, 0x28, 0x65, 0xe0, 0x62, 0x7c, 0x9b, 0x41, 0x6f, 0x12, 0xc3, 0xac,
+        0x78, 0xe4, 0xfc, 0x6b, 0x7b, 0x0a, 0xb4, 0x50, 0xd6, 0xf2, 0x45, 0xf7, 0x93,
+        0x04, 0xaf, 0xf2, 0xb7, 0x26, 0x94, 0xee, 0xad, 0x92, 0x05, 0x6d, 0xe5, 0xc6,
+        0xd6, 0x36, 0xdc, 0xa5, 0x69, 0x0f, 0xc8, 0x99, 0xf2, 0x1c, 0x4e, 0x0d,
+    };
+
+    std::copy(std::begin(mp_cert2), std::end(mp_cert2), reg_data.cert_data);
+
+    if (0x01 != reg_data.cert_enable)
+    {
+      PanicAlert("M+ Failure! Game requested cert2 with value other than 0x01. M+ will disconnect "
+                 "shortly unfortunately. Reconnect wiimote and hope for the best.");
+    }
+
+    // A real wiimote takes about 2 seconds to reach this state:
+    reg_data.cert_ready = 0x1a;
+    INFO_LOG(WIIMOTE, "M+ cert 2 ready!");
+  }
+
+  // TODO: make sure a motion plus report is sent first after init
+
+  // On real mplus:
+  // For some reason the first read seems to have garbage data
+  // is_mp_data and extension_connected are set, but the data is junk
+  // it does seem to have some sort of pattern though, byte 5 is always 2
+  // something like: d5, b0, 4e, 6e, fc, 2
+  // When a passthrough mode is set:
+  // the second read is valid mplus data, which then triggers a read from the extension
+  // the third read is finally extension data
+  // If an extension is not attached the data is always mplus data
+  // even when passthrough is enabled
+
+  // Real M+ seems to only ever read 6 bytes from the extension.
+  // Data after 6 bytes seems to be zero-filled.
+  // After reading, the real M+ uses that data for the next frame.
+  // But we are going to use it for the current frame instead.
+  constexpr int EXT_AMT = 6;
+  // Always read from 0x52 @ 0x00:
+  constexpr u8 EXT_SLAVE = ExtensionPort::REPORT_I2C_SLAVE;
+  constexpr u8 EXT_ADDR = ExtensionPort::REPORT_I2C_ADDR;
+
+  // Try to alternate between M+ and EXT data:
+  auto& mplus_data = *reinterpret_cast<DataFormat*>(data);
```

Comment:
Please don't. This is undefined behavior. Memcpy what you need into the data.

Reply:
I know it's UB. This appears a lot in Dolphin. Could/should we add some kind of BitCastArray function to BitUtils for situations like this? I don't think we should just have memcpys strewn about.

Reply:
Fixed.